Update runtime repo to use Preview 7 SDK

agocke opened the pull request.

agocke committed:
Update runtime repo to use Preview 6 SDK

agocke committed:
Fix json encoding

agocke committed:
Use TrimMode=full instead of TrimmerDefaultAction

agocke committed:
Also set TrimMode=full in TrimmingTests props file

ViktorHofer reviewed (approved):
LGTM. Of course, before merging we need to go through the usual process of communicating the change.

agocke commented:
@MichalStrehovsky @LakshanF  Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?

MichalStrehovsky commented:
> @MichalStrehovsky @LakshanF Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?

Looking... one thing that's missing is adding handling of the new TrimMode to NativeAOT targets. But the failure mode doesn't look like that's the problem.

MichalStrehovsky commented:
> > @MichalStrehovsky @LakshanF Looks like some NativeAOT libraries tests are failing after an SDK update. Ideas?
> 
> Looking... one thing that's missing is adding handling of the new TrimMode to NativeAOT targets. But the failure mode doesn't look like that's the problem.

It's from the new aggresse trimming default in Preview 6. xUnit is calling `PropertyInfo.GetValue` on `CollectionBehaviorAttribute.DisableTestParallelization` but we trimmed the getter (the setter was left because the attribute is applied on something and we need the setter to materialize the `Attribute` instance at runtime - the getter is fair game).

This is actually a very lucky catch. Only a couple of the tests fail. The rest look like this:

```
Running assembly:System.Runtime.Tests, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
Finished System.Runtime.Tests, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

Tests run: 0, Errors: 0, Failures: 0, Skipped: 0. Time: 0.0130819s
```

The fix will be along the lines of adding `<TrimMode>partial</TrimMode>` to `eng\testing\tests.singlefile.targets`.

MichalStrehovsky commented on src/tests/nativeaot/Directory.Build.props:
```diff
@@ -1,7 +1,7 @@
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <TrimmerDefaultAction>link</TrimmerDefaultAction>
+    <TrimMode>full</TrimMode>
```
We'll probably need to do the opposite here - nativeaot tests expect trimming and require the (now default) TrimMode=full. So this line is technically not needed. But the rest of the coreclr test tree will want to keep the partial TrimMode.

MichalStrehovsky commented on src/coreclr/tools/aot/crossgen2/crossgen2.props:
```diff
@@ -14,7 +14,7 @@
     <ServerGarbageCollection>true</ServerGarbageCollection>
     <!-- Trim all dependent assemblies as though they were all marked IsTrimmable. -->
-    <TrimmerDefaultAction>link</TrimmerDefaultAction>
+    <TrimMode>full</TrimMode>
```
Do we need this line if this is now the default?

agocke replied:
No, we don't, I was just being explicit.

agocke commented:
Does NativeAOT understand `TrimMode=partial` or does that need to be added?

agocke committed:
Add TrimmerDefaultAction for NativeAOT library testing

agocke committed:
Remove usage of default trim mode

agocke commented:
/azp run runtime-extra-platforms,runtime-wasm

MichalStrehovsky commented:
> Does NativeAOT understand `TrimMode=partial` or does that need to be added?

I think https://github.com/dotnet/runtime/pull/71117 is what's needed. Could be folded into this PR to avoid adding the TODO.

agocke committed:
Replace with TrimMode partial

agocke commented:
/azp run runtime-extra-platforms,runtime-wasm

agocke committed:
Merge remote-tracking branch 'upstream/main' into runtime-p6

agocke commented:
/azp run runtime-extra-platforms,runtime-wasm

agocke commented:
It looks like the installer tests are broken by https://github.com/dotnet/sdk/issues/26462. The test execution happens through Arcade, so we can't pass DOTNET_ROOT as is.

I see two options for workaround:

1. Add a capability to Arcade to set environment variables for the test process
2. Update to an SDK daily build that contains the fix.

I don't see any way to use the Preview 6 SDK otherwise.

stephentoub commented:
> I think this should be updated to P7 as soon as that's released and we should merge ASAP

Yes.  FYI, we're going to need to update our C# compiler very soon to pick up changes from @cston around scoped, but at least the dotnet-pgo project (and maybe others) will fail to build with that compiler until we move to a newer SDK that has consistency between exposing the ByRefsFeature and having relevant APIs annotated as scoped.
cc: @AaronRobinsonMSFT

ViktorHofer commented:
@dotnet/runtime-infrastructure as @agocke is out, does someone else want to drive this change now that P7 is out?

jkoritzinsky committed:
Merge branch 'runtime-p6' of github.com:agocke/runtime into p7-sdk

jkoritzinsky committed:
Update SDK to preview 7

jkoritzinsky commented:
I can take over on this

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into p7-sdk

jkoritzinsky commented:
We're seeing JIT asserts here in Importation when using the live JIT.

cc: @dotnet/jit-contrib can someone take a look at this?

JulieLeeMSFT commented:
@kunalspathak PTAL.

jakobbotsch commented:
This sounds suspiciously much like https://github.com/dotnet/roslyn/issues/62864.

kunalspathak commented:
Yep, looking at the errors we are hitting

`assert(!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException");`

jkoritzinsky commented:
So I guess we need to update the Roslyn version to at least a version with the fix?

hoyosjs committed:
Update Roslyn to 4.4.0-2.22412.11

AaronRobinsonMSFT commented:
This is an expected issue with the pgo project. I'll push up a fix.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime-p6

AaronRobinsonMSFT committed:
Merge in main and update pgo project

AaronRobinsonMSFT commented:
@MichalStrehovsky or @vitek-karas Any thoughts on the NativeAOT failures?

```
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2070: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2070: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2070: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2070: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.TypeInfo.DeclaredMethods.get'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2067: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2067: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92): Trim analysis error IL2067: ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/System.Runtime.Intrinsics/tests/System.Runtime.Intrinsics.Tests.csproj]
##[error]/_/src/xunit.execution/Extensions/ReflectionAbstractionExtensions.cs(92,0): error IL2067: (NETCORE_ENGINEERING_TELEMETRY=Build) ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethods(Type)'. The parameter 'type' of method 'ReflectionAbstractionExtensions.GetMethodInfoFromIMethodInfo(Type,IMethodInfo)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

AaronRobinsonMSFT committed:
Add ByRefFields feature flag.

radical commented:
/azp run runtime-extra-platforms,runtime-wasm

MichalStrehovsky committed:
Inline deleted ILLink.targets logic

MichalStrehovsky committed:
Cleanups

MichalStrehovsky commented:
> @MichalStrehovsky or @vitek-karas Any thoughts on the NativeAOT failures?

This is WarnAsError in action.

Apparently we deleted what it means to SuppressTrimAnalysisWarnings in https://github.com/dotnet/linker/pull/2930 and NativeAOT no longer suppresses these. I've inlined what that means into the targets to work around and filed a 7.0 issue to fix it because this is customer-visible (https://github.com/dotnet/runtime/issues/73926).

I'm puzzled why we're seeing it here because the break merged into linker repo 10 days ago and there's no way that made it into Preview 7 (in fact, troubleshooting this took longer than I'm willing to admit because the .targets in P7 SDK still have this suppression but we're picking it up from a NuGet instead).

MichalStrehovsky commented:
The Linux_musl x64 Debug failure is https://github.com/dotnet/runtime/issues/73930 (a test touched 4 hours ago, so I'm going with the "unrelated to what we're doing here" theory).

We're good to merge once llvmfullaot is done with its thing.

teo-tsirpanis commented on src/coreclr/tools/dotnet-pgo/SPGO/LbrEntry.cs:
```diff
@@ -74,7 +74,7 @@ public unsafe struct LbrTraceEventData64
         {
             IntPtr entriesOffset = Unsafe.ByteOffset(ref Unsafe.As<LbrTraceEventData64, byte>(ref this), ref Unsafe.As<LbrEntry64, byte>(ref _entries));
-            return MemoryMarshal.CreateSpan(ref _entries, (totalSize - (int)entriesOffset) / sizeof(LbrEntry64));
+            return MemoryMarshal.CreateSpan(ref Unsafe.AsRef<LbrEntry64>(Unsafe.AsPointer(ref _entries)), (totalSize - (int)entriesOffset) / sizeof(LbrEntry64));
```
Won't that cause a GC hole? Same with above.

MichalStrehovsky replied:
Good catch. We either need to make this a `ref struct` with a big "DO NOT CLEAN UP THE REF PART" comment, or use `fixed` to grab the pointer instead.

jakobbotsch replied:
Yes it will. I'll push a fix.

jakobbotsch replied:
> We either need to make this a ref struct with a big "DO NOT CLEAN UP THE REF PART" comment, or use fixed to grab the pointer instead.

I think we can make these static functions that take the owning struct by ref.

teo-tsirpanis replied:
Can't we make it

```suggestion
            return MemoryMarshal.CreateSpan(ref _entries, (totalSize - (int)entriesOffset) / sizeof(LbrEntry64));
```

and add `[UnscopedRef]` to the method?

jakobbotsch replied:
Yes we can, that's an even smaller changeset, I'll do that. Thanks for the suggestion.

AaronRobinsonMSFT replied:
This isn't a GC hole in this case. It is actually correct.

AaronRobinsonMSFT replied:
We were trying to keep this narrow due to all the changes in flight. The problem here is the project is referencing reference assemblies from Preview 7 to build and the annotations here are simply wrong. Once the repo moves to RC1, this can be removed and updated appropriately.

AaronRobinsonMSFT replied:
i shouldn't say "correct", rather it is "safe.

AaronRobinsonMSFT replied:
I really wish this hadn't been touched.

jakobbotsch replied:
@AaronRobinsonMSFT I own this code -- I would prefer to have the code be correct by construction. I agree that it was fine before, but it was leaving an implicit assumption on how I am going to work with this code in the future.

AaronRobinsonMSFT replied:
I see. This was something I was shepherding through for `ref` field support so it was in a state I knew to track. I will tag you if anything more comes up.

jakobbotsch replied:
Ah okay, sorry about that, I did not mean to step on your toes here. 

> I will tag you if anything more comes up.

Thanks, I appreciate it.

AaronRobinsonMSFT replied:
Since I've also got the owner here, can we get this moved to build against the live build? Building against the repo's SDK is the root issue here.

jakobbotsch replied:
I can look into it, I'm not totally sure on the constraints about that (we are publishing dotnet-pgo as a nuget package, would that be affected?). I probably need to involve @davidwrighton here.
I assume it can be done separately?

AaronRobinsonMSFT replied:
> I assume it can be done separately?

Absolutely. My only concern is this project is the only one I found that seems to use the repo's SDK and could be heavily impacted by complex language changes that force updates to the public API surface. This is a longer-term concern and just something to keep in mind as the language evolves to have more of these low-level friendly features.

MichalStrehovsky replied:
> My only concern is this project is the only one I found that seems to use the repo's SDK and could be heavily impacted by complex language changes that force updates to the public API surface

Everything under src/coreclr/tools builds against the repo's SDK. Building against live bits has build sequencing issues where one wouldn't be able to build crossgen2 without building libs first, or run crossgen2 without building the host.

If we want to switch to live, I would want to know what the VS debug experience for live built things looks like. I would still prefer we don't do that because I kind of like I don't need to build coreclr VM, host, or other stuff for NativeAOT development.

AaronRobinsonMSFT replied:
> Building against live bits has build sequencing issues

Ah. Thanks @MichalStrehovsky. This is an important consideration to note. Perhaps this wouldn't justify the resulting dev-loop pain.

jakobbotsch committed:
Ensure no GC hole for LbrTraceEventData

There was no GC hole here in practice since the users have unmanaged
pointers. However, the better fix is to just mark it as UnscopedRef.

radical commented:
/azp run runtime-wasm

radical commented:
@jakobbotsch build failed

jakobbotsch commented:
> @jakobbotsch build failed

Hmm strange, it builds fine from within VS locally. I suppose we cannot use this attribute yet in the project. Let me go back to my other fix.

jakobbotsch committed:
Use static workaround since UnscopedRef is not yet available

radical commented:
/azp run runtime-wasm

radical commented:
The wasm - AOT/EAT failures are present on `main` too, and are - https://github.com/dotnet/runtime/issues/73961 .
The wasm debugger test failure is - https://github.com/dotnet/runtime/issues/73957

jkoritzinsky commented:
Linux x64 Debug leg failure is https://github.com/dotnet/runtime/issues/73930

jkoritzinsky commented:
All failures are accounted for. Merging this in.

ViktorHofer commented:
Thanks for getting this in everyone. @jkoritzinsky as usual, can you please send out the notice to the team including instructions on how to obtain the new SDK (for global installs)?

jkoritzinsky commented:
Will do!
